### PR TITLE
Notifier interface, generic webhooks and slack webhooks

### PR DESCRIPTION
**Pull request by j4k**
## Description
- Related to issue #214
- Add a generic webhook to push updates to flags to a user specified endpoint
- Add a slack specific webhook to push updates to a slack channel like this:

Screenshot 2019-03-19 at 12 10 04

Screenshot 2019-03-19 at 12 10 44

## Motivation and Context
We needed notifications around who was doing what with the flags we had active, when they were enabled / disabled etc

## How Has This Been Tested?
Manually and with automated tests

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by zhouzhuojie** [comment]
Thanks, this is awesome! I will take a look soon

**Comment by j4k** [comment]
thanks, I will fix up

File: pkg/notify/notify.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	f.Preload(db)
+
+	for _, integration := range integrations {
+		err := integration.notifier.Notify(f, b, i)
+
+		if err != nil {
+			logrus.WithFields(logrus.Fields{
+				"err":         err,
+				"flagID":      f.ID,
+				"flagStatus":  f.Enabled,
```
[the remark this answers is not in the document]
Comment (j4k) [reply pkg/notify/notify.go]: just so theres another reference to it

File: pkg/notify/slack.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+	}
+	if str, ok := i.(itemType); ok {
+		return strings.Title(strings.ToLower(string(str)))
+	}
+	return strings.Title(strings.ToLower(""))
+}
+
+// buildSlackRequest builds the correct slack req ready for sending, based on the event in question
+func buildSlackRequest(f *entity.Flag, b notify, i itemType) *slackReq {
+	var blocks []block
```
[the remark this answers is not in the document]
Comment (j4k) [reply pkg/notify/slack.go]: I did have a go with the templating bits in go but couldn't really come up with something super clean, mind chucking up an example or having a hack on it yourself? I know this is not pretty.

**Comment by j4k** [comment]
@zhouzhuojie - have thrown retryable http into the mix as well as timeouts and everything else you've suggested bar the templates, mind giving some input to the direction?

File: pkg/notify/notify.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	"github.com/checkr/flagr/pkg/config"
+	"github.com/checkr/flagr/pkg/entity"
+	"github.com/jinzhu/gorm"
+	"github.com/sirupsen/logrus"
+)
+
+// A Notifier notifies about flags state
+// It returns an error if unsuccessful
+type Notifier interface {
+	Notify(*entity.Flag, itemAction, itemType) error
```
Comment (zhouzhuojie) [inline pkg/notify/notify.go]: I also want to know your preference of what to include in the `Notify` interface. For example, what do you think of 
- including the diff of current vs previous flag_snapshot? Sometimes, the flag itself can include many many segments, but the diff can be small enough and easy to visualize in Slack.
- including `updated_by`?

**Comment by j4k** [comment]
Sorry for the delayed response I have been away.

I'll get this stuff fixed up this week!

**Comment by pacoguzman** [comment]
@j4k sorry to bother you, are you going to be able to complete this? this is something that we can use too

**Comment by vic3lord** [comment]
I was about to add the same functionality to flagr so awesome work!
Is there a way to also notify on a flag that's rolled out 100% for x time straight (1 week) so we know it's ready and can clean the code? this would be really helpful

**Comment by borjafv** [comment]
I'm also really interesed in this feature, is there something we can do to help out to give it the final push to master? Thanks!

**Comment by pacoguzman** [comment]
We started to work on this, and we're continuing this [here](https://github.com/checkr/flagr/compare/master...pacoguzman:generic-notifier), so I'm solving the last requests you made before, and I'll we can discuss pending feedback in the meantime

File: pkg/notify/notify.go (hunk 3 above)
Comment (pacoguzman) [reply pkg/notify/notify.go:13]: I think `updated_by` makes a lot of sense to provide it. About the diff I would say it makes sense too so what do you think on providing all of the data so notifiers can make us of it as they wish?

File: pkg/notify/client.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	}
+
+	if resp.StatusCode == 0 || resp.StatusCode >= 500 {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// Backoff specifies a policy on how frequently we should retry
```
[the remark this answers is not in the document]
Comment (pacoguzman) [reply pkg/notify/client.go:34]: I'll try to adapt it to make use of the existing retry package and keep you posted

**Pull request closed without merging**